website/project/decorators.py

```diff
@@ -183,6 +186,22 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def check_can_download_preprint_file(user, node):
+    """View helper that returns whether a given user can download unpublished preprint files.
+
+    :rtype: boolean
+    """
+    if not node.preprints.exists():
+        return False
+
+    preprint = node.preprints.filter(guids___id=request.base_url.split('/')[-2]).first()
+    provider = preprint.provider
+
+    is_admin_or_moderator = bool(user.groups.exclude(name__startswith='collections_') & Group.objects.filter(
+        Q(name='reviews_{}_moderator'.format(provider._id)) | Q(name='reviews_{}_admin'.format(provider._id))))
+
+    return (provider.is_reviewed and is_admin_or_moderator) or user.has_perm('view_submissions', provider)
```

Comment:
Two questions:
1. Is this ever actually hit? `check_can_access` is only used by `check_contributor_auth`, which is used once by addons (completely unrelated) and by the `must_be_contributor`-type decorators. As far as I'm aware, the only case affected by the `download` kwarg is `resolve_guid_download`, which is unaffected by said decorators.
2. If this was hit, wouldn't it block access to published preprints for... everyone other than a moderator?

Reply:
It hits. Here https://github.com/CenterForOpenScience/osf.io/blob/0fab45b035e0a224d7e057a9354ae9a553761c8c/website/views.py#L309, the file deep url matches `/project/<pid>/files/<provider>/<path:path>/` which maps to `addon_view_or_download_file` https://github.com/CenterForOpenScience/osf.io/blob/develop/addons/base/views.py#L624

Reply:
Ah, okay. The first half of the conditional (and therefore the query above it) is unnecessary, though. That permissions only assigned to the admin and moderator django-guardian groups for each provider -- you should let that library do the heavy lifting.